File: app/frontend/src/core/interfaces/types/npc/ReputationSystem.ts
import { InteractionResult } from '../../systems/npc/InteractionSystem';

export interface ReputationChange {
  value: number;
  reason: string;
  timestamp: number;
}

export interface ReputationSystem {
  processInteractionReputation(
    npcId: string,
    targetId: string,
    result: InteractionResult
  ): Promise<number>;
  calculateBaseReputation(result: InteractionResult): number;
  applyInteractionModifiers(baseChange: number, result: InteractionResult): number;
  recordReputationChange(
    npcId: string,
    targetId: string,
    value: number,
    result: InteractionResult
  ): void;
  generateChangeReason(result: InteractionResult): string;
  getReputationHistory(npcId: string, targetId: string): ReputationChange[];
  getAggregateReputation(npcId: string, targetId: string): number;
  updateReputation(
    npcId: string,
    change: number,
    context: any
  ): Promise<void>;
  getReputation(
    npcId: string,
    groupId?: string
  ): Promise<number>;
  calculateReputationChange(
    baseChange: number,
    context: any
  ): number;
  decayReputation(
    npcId: string,
    timePassed: number
  ): Promise<void>;
  getReputationHistory(
    npcId: string,
    timeRange: {
      start: number;
      end: number;
    }
  ): Promise<{
    timestamp: number;
    change: number;
    reason: string;
  }[]>;
} 

File: backend/app/02_medium_complexity/src/core/interfaces/types/npc/ReputationSystem.py
from typing import Any, Dict, List



class ReputationChange:
    value: float
    reason: str
    timestamp: float
class ReputationSystem:
    processInteractionReputation(
    npcId: str,
    targetId: str,
    result: InteractionResult
  ): Awaitable[float>
    calculateBaseReputation(result: InteractionResult): float
    applyInteractionModifiers(baseChange: float, result: InteractionResult): float
    recordReputationChange(
    npcId: str,
    targetId: str,
    value: float,
    result: InteractionResult
  ): None
    generateChangeReason(result: InteractionResult): str
    getReputationHistory(npcId: List[str, targetId: str): ReputationChange]
    getAggregateReputation(npcId: str, targetId: str): float
    updateReputation(
    npcId: str,
    change: float,
    context: Any
  ): Awaitable[None>
    getReputation(
    npcId: str,
    groupId?: str
  ): Awaitable[float>
    calculateReputationChange(
    baseChange: float,
    context: Any
  ): float
    decayReputation(
    npcId: str,
    timePassed: float
  ): Awaitable[None>
    getReputationHistory(
    npcId: str,
    timeRange: Dict[str, Any][]>
} 